lib/handler/redirect.js

```diff
@@ -54,6 +54,13 @@ class RedirectHandler {
       return this.handler.onHeaders(statusCode, headers, resume)
     }
 
+    // Add the URL to the list of URLs visited while following redirections
+    if (!Array.isArray(this.handler.context.history)) {
+      this.handler.context.history = []
+    }
+
+    this.handler.context.history.push(new URL(this.opts.path, this.opts.origin).toString())
```

Comment:
What if handler doesn't have context?

Reply:
Can we pass context similar to how we pass `opaque`? also make it optional.

Reply:
I'd prefer if we don't have properties on handlers in terms of user api.

Reply:
Ok, let me see what I can do. It's also a good idea because it enables custom handlers to be customizable without requiring changes in undici internals. 

I'll just clarify in docs that the context object might be modified by undici. Does it sound reasonable?

Reply:
> I'll just clarify in docs that the context object might be modified by undici. 

Let's try it and see.

Reply:
Another alternative could be to have another method on the handler:

```js
handler.onContext({ history })
```

Reply:
I like it. It requires less modification from the current code and does not require a change in the external API. 
Also, handler implementers can better choose how to handle the context they receive (like "replacing" instead of "merging" and so on.

Stay tuned.

Reply:
Nevermind. I just realized that having a method, especially in the redirect handler would be messy since it would need another method/property to get the current history.

I'll go with user provided object.

Reply:
Why? The redirect handler creates the property on itself and then just calls `onContext` to give the user access to it.

Reply:
> I'll go with user provided object.

Probably also fine. We can start with trying that.

Reply:
> Why? The redirect handler creates the property on itself and then just calls onContext to give the user access to it.

Because the RedirectHandler is recreated every time (decreasing maxRedirections) when following the chain. I would then store the history in opts and then call `onContext`. At that point is easier to use the user object.

Reply:
So, I have analyzed both solutions.

The first one, passing the context like in opaque is not working as the opts are lost in the API handler (since we destructure there). Saving the context there is not different from what I'm doing now and it complicates the external API so I gave up on that.

The `onContext` approach instead has the issue that if the chain is interrupted (if maxRedirections reaches 0 and 3xx is still  returned) is not called at all as it is inside RedirectHandler and not inside the Agent.

I also tried to call `onContext` expecting the bottom handler to return an object to modify. But this implies saving the context in the bottom handler which is what I'm doing at the moment. 

I also tried to mix the approaches, no luck.

I therefore think we should keep the current implementation, unless you have better idea.

Reply:
Thanks for putting the effort into considering both approaches. I'll also have a go at this.

Reply:
Ok, keep me posted then. In the meanwhile I'll just send one more commit that checks that the handler has a context before attempting to set it.